src/payment/models/exchange/exchange-utility.service.ts

```diff
@@ -0,0 +1,49 @@
+import { Injectable } from '@nestjs/common';
+import { BinanceService } from './binance.service';
+import { BitstampService } from './bitstamp.service';
+import { Price } from './dto/price.dto';
+import { KrakenService } from './kraken.service';
+
+type PriceSource = string;
+
+@Injectable()
+export class ExchangeUtilityService {
+  constructor(
+    private readonly krakenService: KrakenService,
+    private readonly binanceService: BinanceService,
+    private readonly bitstampService: BitstampService,
+  ) {}
+
+  async getMatchingPrice(fromCurrency: string, toCurrency: string, matchThreshold = 0.02): Promise<Price> {
+    const mainPrice = await this.krakenService.getPrice(fromCurrency, toCurrency);
+    const [refPrice, refSource] = await this.getReferencePrice(fromCurrency, toCurrency);
+
+    const { price: _mainPrice } = mainPrice;
+    const { price: _refPrice } = refPrice;
+
+    if (Math.abs(_refPrice - _mainPrice) / _mainPrice > matchThreshold)
+      throw new Error(
+        `${fromCurrency} to ${toCurrency} price mismatch (kraken: ${_mainPrice}, ${refSource}: ${_refPrice})`,
+      );
+
+    return mainPrice;
+  }
+
+  private async getReferencePrice(fromCurrency: string, toCurrency: string): Promise<[Price, PriceSource]> {
+    try {
+      return [await this.binanceService.getPrice(fromCurrency, toCurrency), 'binance'];
+    } catch {
+      console.log(`Could not find reference price at Binance. From ${fromCurrency} to ${toCurrency}`);
+    }
+
+    try {
+      return [await this.bitstampService.getPrice(fromCurrency, toCurrency), 'bitstamp'];
+    } catch {
+      console.log(`Could not find reference price at Bitstamp. From ${fromCurrency} to ${toCurrency}`);
+    }
```

Comment:
hmm, where else would you have this info? Buy process logs are buy process related and it would be quite hard to tell where potential issue with prices comes from

Reply:
For USDC you will never get the price from Binance, for every processed batch you will get the upper log. So the question is, what do you gain from these logs?

Reply:
aha, understood now, thanks! makes sense